packages/medusa/src/interfaces/payment-processor.ts

```diff
@@ -0,0 +1,215 @@
+  error: string
+  code: number
+  details: any
+}
+
+/**
+ * The new payment service plugin interface
+ * This work is still experimental and can be changed until it becomes stable
+ */
+export interface PaymentProcessor {
+  /**
+   * Return a unique identifier to retrieve the payment plugin provider
+   */
+  getIdentifier(): string
+
+  /**
+   * Used to initialise anything like an SDK or similar
+   */
+  init(): Promise<void>
+
+  /**
+   * Initiate a payment session with the external provider
+   */
+  initiatePayment(
+    context: PaymentProcessorContext
+  ): Promise<PaymentProcessorError | PaymentProcessorSessionResponse>
+
+  /**
+   * Update an existing payment session
+   * @param paymentId
+   * @param context
+   */
+  updatePayment(
+    paymentId: string,
+    context: PaymentProcessorContext
+  ): Promise<PaymentProcessorError | void>
+
+  /**
+   * Refund an existing session
+   * @param paymentId
+   * @param context
+   */
+  refundPayment(
+    paymentId: string,
+    context: PaymentProcessorContext
+  ): Promise<PaymentProcessorError | void>
+
+  /**
+   * Authorize an existing session if it is not already authorized
+   * @param paymentId
+   * @param context
+   */
+  authorizePayment(
+    paymentId: string,
+    context: PaymentProcessorContext
+  ): Promise<PaymentProcessorError | void>
+
+  /**
+   * Capture an existing session
+   * @param paymentId
+   * @param context
+   */
+  capturePayment(
+    paymentId: string,
+    context: PaymentProcessorContext
+  ): Promise<PaymentProcessorError | void>
+
+  /**
+   * Delete an existing session
+   * @param paymentId
+   */
+  deletePayment(paymentId: string): Promise<PaymentProcessorError | void>
+
+  /**
+   * Retrieve an existing session
+   * @param paymentId
+   */
+  retrievePayment(
+    paymentId: string
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  >
+
+  /**
+   * Cancel an existing session
+   * @param paymentId
+   */
+  cancelPayment(paymentId: string): Promise<PaymentProcessorError | void>
+
+  /**
+   * Return the status of the session
+   * @param paymentId
+   */
+  getPaymentStatus(paymentId: string): Promise<PaymentSessionStatus>
+
+  /**
+   * An optional method to implement in order to retrieve the methods saved for a customer in case the provider
+   * provide such feature. It is based on the collected data that have been given by the plugin
+   * and saved on the customer
+   *
+   * @example
+   * getSavedMethods(
+   *   customer: Customer
+   * ): Promise<Record<string, unknown>[]> {
+   *  if (customer.metadata?.stripe_id) {
+   *    const methods = await this.stripe_.paymentMethods.list({
+   *      customer: customer.metadata.stripe_id
+   *      type: "card",
+   *    })
+   *
+   *    return methods.data
+   *  }
+   *
+   *  return []
+   * }
+   *
+   * @param customer
+   */
+  getSavedMethods(customer: Customer): Promise<Record<string, unknown>[]>
```

Comment:
thought: think we should kill this in the first iteration.

Reply:
Yeah probably, it means that the get-payment-method end point will not be effective for a payment processor. Is that ok?